### PR TITLE
AEA-2143 Handle Clinical Test-Packs

### PR DESCRIPTION
**Pull request by booshibooshi**

### CODE REVIEW

File: tool/site/client/src/pages/loadPage.tsx [hunk 1]
```diff
@@ -65,16 +65,17 @@ const LoadPage: React.FC = () => {
...
-        for (let i = 0; i < bundles.length; i += batchSize) {
-          const batch = bundles.slice(i, i + batchSize)
-          const response = await axiosInstance.post<LoadResponse>(`${baseUrl}prescribe/edit`, batch)
+        const chunkSize = 10
+        for (let i = 0; i < bundles.length; i += chunkSize) {
+          const chunk = bundles.slice(i, i + chunkSize)
+          const response = await axiosInstance.post<LoadResponse>(`${baseUrl}prescribe/edit`, chunk)
           loadResponses.push(getResponseDataIfValid(response, isLoadResponse))
         }
         return loadResponses
```
Comment (JackRobinsonIW) [inline tool/site/client/src/pages/loadPage.tsx:81]: ignore this diff

File: tool/site/client/src/requests/axiosInstance.ts [hunk 2]
```diff
@@ -3,3 +3,10 @@ import axios from "axios"
 export const axiosInstance = axios.create({
   validateStatus: () => true
 })
+
+// todo: refresh token if required
+axiosInstance.interceptors.request.use(function (config) {
+  return config
+}, function (error) {
+  return Promise.reject(error)
+})
```
Comment (JackRobinsonIW) [inline tool/site/client/src/requests/axiosInstance.ts:12]: remove this

File: tool/site/client/tests/pages/sendPostSignPage.spec.tsx [hunk 3]
```diff
@@ -20,37 +19,6 @@ beforeEach(() => moxios.install(axiosInstance))
...
-  const container = await renderPage()
-
-  expect(screen.getByText(prescriptionId)).toBeTruthy()
-  expect(screen.getByText(JSON.stringify("JSON Request"))).toBeTruthy()
-  expect(screen.getByText("XML Request")).toBeTruthy()
-  expect(screen.getByText(JSON.stringify("JSON Response"))).toBeTruthy()
-  expect(screen.getByText("XML Response")).toBeTruthy()
-  expect(pretty(container.innerHTML)).toMatchSnapshot()
-})
-
```
Comment (JackRobinsonIW) [inline tool/site/client/tests/pages/sendPostSignPage.spec.tsx:53]: want to re-add this alongside the single prescription user journey

File: tool/e2e-tests/helpers.ts [hunk 4]
```diff
@@ -164,6 +165,43 @@ export async function claimAmendPrescriptionUserJourney(
...
+  await setMockSigningConfig(driver)
+  await createPrescription(driver)
+  await loadPredefinedExamplePrescription(driver)
+  await editPrescriptionOrganisation(driver, newOrganisation)
+  await sendPrescription(driver)
+  await checkApiResult(driver)
+  return await getCreatedPrescriptionId(driver)
+}
+
+export async function editPrescriptionOrganisation(
```
Comment (callumtipper-IW) [inline tool/e2e-tests/helpers.ts:182]: Possibly worth tidying up where the test functions are, in the test file if they're not shared, in helpers if they are?

File: tool/e2e-tests/prescribe/sendPrescriptionsFromTestPack.ts [hunk 5]
```diff
@@ -3,7 +3,7 @@ import {sendBulkPrescriptionUserJourney} from "../helpers"
 import {loadClinicalTestPack1Examples} from "../test-packs/test-packs"
 
 describe("firefox", () => {
-  test("can send prescriptions from test pack 1", async () => {
-    await sendBulkPrescriptionUserJourney(driver, loadClinicalTestPack1Examples, 5)
+  test("can send prescriptions from clinical test pack 1", async () => {
+    await sendBulkPrescriptionUserJourney(driver, loadClinicalTestPack1Examples, 25)
```
Comment (callumtipper-IW) [inline tool/e2e-tests/prescribe/sendPrescriptionsFromTestPack.ts:7]: why 25? the example xlsx used has 214 entries

File: tool/site/client/src/components/send/resultSummaries.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            <Table.Cell>Success</Table.Cell>
+            <Table.Cell/>
+          </Table.Row>
+        </Table.Head>
+        <Table.Body>
+          {sendResult.results.map(result => (
+            <Table.Row key={result.prescription_id}>
+              <Table.Cell>{result.bundle_id}</Table.Cell>
+              <Table.Cell>{result.prescription_id}</Table.Cell>
+              <Table.Cell>{result.success === "unknown" ? <Spinner /> : result.success ? <TickIcon /> : <CrossIcon />}</Table.Cell>
```
Comment (callumtipper-IW) [inline tool/site/client/src/components/send/resultSummaries.tsx:37]: component

File: tool/e2e-tests/helpers.ts (hunk 4 above)
Comment (JackRobinsonIW) [reply tool/e2e-tests/helpers.ts:182]: Moved across

File: tool/site/client/src/components/send/resultDetail.tsx [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import ButtonList from "../common/buttonList"
+
+interface ResultDetailProps {
+  sendResultDetail: SendResultDetail
+}
+
+export const ResultDetail: React.FC<ResultDetailProps> = ({sendResultDetail}) => {
+  return (
+    <>
+      <Label isPageHeading>Send Result {sendResultDetail.success ? <TickIcon/> : <CrossIcon/>}</Label>
```
[the remark this answers is not in the document]
Comment (JackRobinsonIW) [reply tool/site/client/src/components/send/resultDetail.tsx:16]: DOne

File: tool/site/server/src/routes/prescribe/edit.ts [hunk 8]
```diff
@@ -22,28 +22,47 @@ export default [
...
+        return responseToolkit.response({}).code(200)
       }
 
+      const requestPrescriptionIds: Array<{bundleId: string | undefined, prescriptionId: string}> = []
+
       prepareBundles.forEach((prepareBundle: fhir.Bundle) => {
         const prescriptionId = getMedicationRequests(prepareBundle)[0].groupIdentifier?.value ?? ""
         if (prescriptionId) {
-          prescriptionIds.push(prescriptionId)
+          const prescription = {bundleId: prepareBundle.id, prescriptionId}
```
Comment (callumtipper-IW) [inline tool/site/server/src/routes/prescribe/edit.ts:38]: inline type for this object

File: tool/site/server/src/routes/state/session.ts [hunk 9]
```diff
@@ -35,7 +35,7 @@ export default [
     method: "GET",
     path: "/prescriptions",
     handler: async (request: Hapi.Request, h: Hapi.ResponseToolkit): Promise<Hapi.ResponseObject> => {
-      const editingPrescriptionIds = getSessionValue("prescription_ids", request)
+      const editingPrescriptionIds = getSessionValue("prescription_ids", request).map((id: { prescriptionId: string }) => id.prescriptionId)
```
Comment (callumtipper-IW) [inline tool/site/server/src/routes/state/session.ts:38]: using this line a lot in different places, extract?

File: tool/e2e-tests/prescribe/sendPrescriptionsFromTestPack.ts (hunk 5 above)
Comment (JackRobinsonIW) [reply tool/e2e-tests/prescribe/sendPrescriptionsFromTestPack.ts:7]: The signature requests/responses are made in batches of 25, so this test waits for the first 25 to come back successfully (25 ticks on screen) and then moves on

File: tool/site/client/src/components/send/resultSummaries.tsx (hunk 6 above)
Comment (JackRobinsonIW) [reply tool/site/client/src/components/send/resultSummaries.tsx:37]: Done

File: tool/site/server/src/routes/prescribe/edit.ts (hunk 8 above)
Comment (JackRobinsonIW) [reply tool/site/server/src/routes/prescribe/edit.ts:38]: Done

File: tool/site/server/src/services/communication/mock-signing-client.ts [hunk 10]
```diff
@@ -21,7 +21,7 @@ export class MockSigningClient implements SigningClient {
 
   async makeSignatureDownloadRequest(): Promise<any> {
     const mockCertificate = ""
-    const mockSignatures = getSessionValue("prescription_ids", this.request).map((id: string) => {
+    const mockSignatures = getSessionValue("prescription_ids", this.request).map((id: { prescriptionId: string }) => id.prescriptionId).map((id: string) => {
```
[the remark this answers is not in the document]
Comment (JackRobinsonIW) [reply tool/site/server/src/services/communication/mock-signing-client.ts:24]: Fixed by the above util refactor

File: tool/site/client/src/pages/sendPage.tsx [hunk 11]
```diff
@@ -33,10 +33,12 @@ const SendPage: React.FC<SendPageProps> = ({
         }
         const pendingSendResults = sendResultState.results.filter(r => r.success === "unknown")
         if (pendingSendResults.length) {
-          const deltaResult = await sendNextPrescriptionBatch(baseUrl, token, pendingSendResults)
-          const previousResult = sendResultState.results.filter(r => !deltaResult.results.map(r => r.prescription_id).includes(r.prescription_id))
+          const deltaResults = (await sendNextPrescriptionBatch(baseUrl, token, pendingSendResults)).results
+          const detlaResultPrescriptionIds = deltaResults.map(result => result.prescription_id)
+
+          const previousResults = sendResultState.results.filter(result => !detlaResultPrescriptionIds.includes(result.prescription_id))
```
Comment (callumtipper-IW) [inline tool/site/client/src/pages/sendPage.tsx:39]: det-la

**Review by callumtipper-IW — APPROVED** [review]

**Pull request merged**